elyra/contents/parser.py

```diff
@@ -0,0 +1,193 @@
+                for match in regex.finditer(line):
+                    matches.append((key, match))
+        return matches
+
+
+class PythonScriptParser(ScriptParser):
+    def search_expressions(self) -> Dict[str, List]:
+        # TODO: add more key:list-of-regex pairs to parse for additional resources
+        regex_dict = dict()
+
+        # First regex matches envvar assignments of form os.environ["name"] = value w or w/o value provided
+        # Second regex matches envvar assignments that use os.getenv("name", "value") with ow w/o default provided
+        # Third regex matches envvar assignments that use os.environ.get("name", "value") with or w/o default provided
+        # Both name and value are captured if possible
+        envs = [r"os\.environ\[[\"']([a-zA-Z_]+[A-Za-z0-9_]*)[\"']\](?:\s*=(?:\s*[\"'](.[^\"']*)?[\"'])?)*",
+                r"os\.getenv\([\"']([a-zA-Z_]+[A-Za-z0-9_]*)[\"'](?:\s*\,\s*[\"'](.[^\"']*)?[\"'])?",
+                r"os\.environ\.get\([\"']([a-zA-Z_]+[A-Za-z0-9_]*)[\"'](?:\s*\,(?:\s*[\"'](.[^\"']*)?[\"'])?)*"]
+        regex_dict["env_list"] = envs
+        return regex_dict
+
+
+class RScriptParser(ScriptParser):
+    def search_expressions(self) -> Dict[str, List]:
+        # TODO: add more key:list-of-regex pairs to parse for additional resources
+        regex_dict = dict()
+
+        # Tests for matches of the form Sys.setenv("key" = "value")
+        envs = [r"Sys\.setenv\([\"']*([a-zA-Z_]+[A-Za-z0-9_]*)[\"']*\s*=\s*[\"']*(.[^\"']*)?[\"']*\)",
+                r"Sys\.getenv\([\"']*([a-zA-Z_]+[A-Za-z0-9_]*)[\"']*\)(.)*"]
+        regex_dict["env_list"] = envs
+        return regex_dict
+
+
+class ContentParser(LoggingConfigurable):
+    parsers = {
+        'python': PythonScriptParser(),
+        'r': RScriptParser()
+    }
+
+    def parse(self, filepath: str) -> dict:
+        """Returns a model dictionary of all the regex matches for each key in the regex dictionary"""
+
+        properties = {"env_list": {}, "inputs": {}, "outputs": {}}
+        reader = self.get_reader(filepath)
+        parser = self.get_parser(reader.language)
+
+        if not parser:
+            return properties
+
+        for chunk in reader.read_next_code_chunk():
+            if chunk:
+                for line in chunk:
+                    matches = parser.parse_environment_variables(line)
+                    for key, match in matches:
+                        properties[key][match.group(1)] = match.group(2)
+
+        return properties
+
+    def get_reader(self, filepath: str):
+        """
+        Find the proper reader based on the file extension
+        """
+        file_extension = os.path.splitext(filepath)[-1]
+
+        if file_extension == '.ipynb':
+            return NotebookReader(filepath)
+        elif file_extension in ['.py', '.r']:
+            return FileReader(filepath)
+        else:
+            raise ValueError(f'File type {file_extension} is not supported.')
```

Comment:
The check for whether the file exists or is a directory should be moved out of the `FileReader` constructor, otherwise attempts to access non-existent files (whose suffixes are not supported) or directories - will result in the return of an empty result set, when 404 or 400 errors should be returned.

Reply:
Good catch. I'm thinking move `validate_filepath()` to be a `ContentParser` method and call it in `get_reader()` before the file extension if/else block?

Reply:
That will be good.  I think preserving its underscore prefix would be correct as well.  Should probably also add underscores prior to `get_reader` and `get_parser` since they shouldn't be called from consumers.

Reply:
Will do!